stint/private/datatypes.nim

```diff
@@ -10,89 +10,147 @@
+# similar to a binary tree of power of two unsigned integers
+# with "high" and "low" words:
+#
+#                              Stuint[256]
+#            hi: Stuint[128]                  lo: Stuint[128]
+#     hihi: uint64    hilo: uint64    lohi: uint64    lolo: uint64
+#
+# This follows paper https://hal.archives-ouvertes.fr/hal-00582593v2
+# "Recursive double-size fixed precision arithmetic" from Jul. 2016
+# to implement an efficient fixed precision bigint for embedded devices, especially FPGAs.
+#
+# For testing purpose, the flag `-d:stint_test` can be passed at compile-time
+# to switch the backend to uint32.
+# In the future the default backend will become uint128 on supporting compilers.
+#
+# This has following benefits:
+#   - BigEndian/LittleEndian support is trivial.
+#   - Not having for loops help the compiler producing the most efficient instructions
+#     like ADC (Add with Carry)
+#   - Proving that the recursive structure works at depth 64 for uint32 backend means that
+#     it would work at depth 128 for uint64 backend.
+#     We can easily choose a uint16 or uint8 backend as well.
+#   - Due to the recursive structure, testing operations when there is:
+#       - no leaves(uint64)
+#       - a root and leaves with no nodes (uint128)
+#       - a root + intermediate nodes + leaves (uint256)
+#     should be enough to ensure they work at all sizes, edge cases included.
+#   - Adding a new backend like uint128 (GCC/Clang) or uint256 (LLVM instrinsics only) is just adding
+#     a new case in the `uintImpl` macro.
+#   - All math implementations of the operations have a straightforward translation
+#     to a high-low structure, including the fastest Karatsuba multiplication
+#     and co-recursive division algorithm by Burnikel and Ziegler.
+#     This makes translating those algorithms into Nim easier compared to an array backend.
+#     It would also probably require less code and would be much easier to audit versus
+#     the math reference papers.
+#   - For implementation of algorithms, there is no issue to take subslices of the memory representation
+#     with a recursive tree structure.
+#     On the other side, returning a `var array[N div 2, uint64]` is problematic at the moment.
+#   - Compile-time computation is possible while due to the previous issue
+#     an array backend would be required to use var openarray[uint64]
+#     i.e. pointers.
+#   - Note that while shift-right and left can easily be done an array of bytes
+#     this would have reduced performance compared to moving 64-bit words.
+#     An efficient implementation on array of words would require checking the shift
+#     versus a half-word to deal with carry-in/out from and to the adjacent words
+#     similar to a recursive implementation.
+#
+# Iterations over the whole integers, for example for `==` is always unrolled.
+# Due to being on the stack, any optimizing compiler should compile that to efficient memcmp
+
+when not defined(stint_test):
```

Comment:
Here I just exchanged the release and test backend so that the release structure is presented first